In [1]:
import pandas as pd

In [2]:
mobile = pd.read_csv("mobile_price_data.csv")

Briefly explore the raw dataset

In [3]:
mobile.head()

mobile_name mobile_price mobile_color  dual_sim  \
0    Realme Narzo 10A (So White, 32 GB)       ₹8,999     So White  Dual Sim   
1     Realme Narzo 10A (So Blue, 32 GB)       ₹8,999      So Blue  Dual Sim   
2     Realme Narzo 10A (So Blue, 64 GB)       ₹9,999      So Blue  Dual Sim   
3    Realme Narzo 10A (So White, 64 GB)       ₹9,999     So White  Dual Sim   
4  Realme Narzo 10 (That Green, 128 GB)      ₹11,999   That Green  Dual Sim   

              disp_size         resolution          os  num_cores mp_speed  \
0  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
1  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
2  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
3  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
4   16.51 cm (6.5 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   

  int_memory   ram                   p_cam              f_cam  \
0      32 GB  3 GB        12MP + 2MP + 2MP   5MP Front Camera   
1      32 GB  3 GB        12MP + 2MP + 2MP   5MP Front Camera   
2      64 GB  4 GB        12MP + 2MP + 2MP   5MP Front Camera   
3      64 GB  4 GB        12MP + 2MP + 2MP   5MP Front Camera   
4     128 GB  4 GB  48MP + 8MP + 2MP + 2MP  16MP Front Camera   

                network bluetooth battery_power mob_width mob_height  \
0  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
1  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
2  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
3  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
4  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh   75.4 mm   164.4 mm   

  mob_depth mob_weight  
0   8.95 mm      195 g  
1   8.95 mm      195 g  
2   8.95 mm      195 g  
3   8.95 mm      195 g  
4      9 mm      199 g

First of all, we remove those units in the columns, by using regex

In [4]:
import re
mobile["disp_size"] = mobile["disp_size"].apply(lambda x: re.sub(r'(.+)(cm)(.+)',r'\1', x))
mobile["mp_speed"] = mobile["mp_speed"].apply(lambda x: re.sub(r'(.+)( GHz)',r'\1', x))
mobile["int_memory"] = mobile["int_memory"].apply(lambda x: re.sub(r'(.+)( GB)',r'\1', x))
mobile["ram"] = mobile["ram"].apply(lambda x: re.sub(r'(.+)( GB)',r'\1', x))
mobile["battery_power"] = mobile["battery_power"].apply(lambda x: re.sub(r'(.+)( mAh)',r'\1', x))
mobile["mob_width"] = mobile["mob_width"].apply(lambda x: re.sub(r'(.+)( mm)',r'\1', x))
mobile["mob_height"] = mobile["mob_height"].apply(lambda x: re.sub(r'(.+)( mm)',r'\1', x))
mobile["mob_depth"] = mobile["mob_depth"].apply(lambda x: re.sub(r'(.+)( mm)',r'\1', x))
mobile["mob_weight"] = mobile["mob_weight"].apply(lambda x: re.sub(r'(.+)( g)',r'\1', x)) 

Splitting "mobile_name" into two columns: "mobile_brand" and "model_name"

In [5]:
mobile[['mobile_brand', 'model_name']] = mobile['mobile_name'].str.split(' ', 1, expand=True)
mobile.drop('mobile_name', inplace=True, axis=1)

mobile["model_name"] = mobile["model_name"].apply(lambda x: re.sub(r'(.*?) \((.*)',r'\1', x))

Splitting "resolution", "f_cam" and "p_cam", for the ease of data analysis

In [6]:
mobile[['resolution_width', 'resolution_height']] = mobile['resolution'].str.split(r'x|×|\*', 1, expand=True)
mobile.drop('resolution', inplace=True, axis=1)
mobile["resolution_height"] = mobile["resolution_height"].astype(str)
mobile["resolution_height"] = mobile["resolution_height"].apply(lambda x: re.sub(r'(\d+)(.*)',r'\1', x))

mobile[['f_cam_1', 'f_cam_2']] = mobile['f_cam'].str.split('+', 1, expand=True)
mobile.drop('f_cam', inplace=True, axis=1)
mobile["f_cam_1"] = mobile["f_cam_1"].astype(str)
mobile["f_cam_2"] = mobile["f_cam_2"].astype(str)
mobile["f_cam_1"] = mobile["f_cam_1"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["f_cam_1"] = mobile["f_cam_1"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["f_cam_1"] = mobile["f_cam_1"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["f_cam_2"] = mobile["f_cam_2"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["f_cam_2"] = mobile["f_cam_2"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["f_cam_2"] = mobile["f_cam_2"].apply(lambda x: re.sub(r'(\D+)',r'', x))

mobile[['p_cam_1', 'p_cam_2', 'p_cam_3', 'p_cam_4']] = mobile['p_cam'].str.split('+', 4, expand=True)
mobile.drop('p_cam', inplace=True, axis=1)
mobile["p_cam_1"] = mobile["p_cam_1"].astype(str)
mobile["p_cam_2"] = mobile["p_cam_2"].astype(str)
mobile["p_cam_3"] = mobile["p_cam_3"].astype(str)
mobile["p_cam_4"] = mobile["p_cam_4"].astype(str)
mobile["p_cam_1"] = mobile["p_cam_1"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_1"] = mobile["p_cam_1"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_1"] = mobile["p_cam_1"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["p_cam_2"] = mobile["p_cam_2"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_2"] = mobile["p_cam_2"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_2"] = mobile["p_cam_2"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["p_cam_3"] = mobile["p_cam_3"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_3"] = mobile["p_cam_3"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_3"] = mobile["p_cam_3"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["p_cam_4"] = mobile["p_cam_4"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_4"] = mobile["p_cam_4"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_4"] = mobile["p_cam_4"].apply(lambda x: re.sub(r'(\D+)',r'', x))

For the "network" column, we just keep the newer and better network type, i.e.: 5G>4G VOLTE>4G>3G>2G

In [7]:
list1 = [] 
for x in mobile["network"]:
    if "5G" in x:
        list1.append("5G")
    elif "4G VOLTE" in x:
        list1.append("4G VOLTE")
    elif "4G" in x:
        list1.append("4G")
    elif "3G" in x:
        list1.append("3G")
    elif "2G" in x:
        list1.append("2G")
mobile["network"] = list1
mobile.rename(columns={'network': 'network_top'}, inplace=True)

Briefly explore our cleaned dataset, to check whether there is any error or mistake during our cleaning process

In [8]:
mobile.head()

mobile_price mobile_color  dual_sim disp_size          os  num_cores  \
0       ₹8,999     So White  Dual Sim    16.56   Android 10  Octa Core   
1       ₹8,999      So Blue  Dual Sim    16.56   Android 10  Octa Core   
2       ₹9,999      So Blue  Dual Sim    16.56   Android 10  Octa Core   
3       ₹9,999     So White  Dual Sim    16.56   Android 10  Octa Core   
4      ₹11,999   That Green  Dual Sim    16.51   Android 10  Octa Core   

  mp_speed int_memory ram network_top  ... mobile_brand model_name  \
0        2         32   3    4G VOLTE  ...       Realme  Narzo 10A   
1        2         32   3    4G VOLTE  ...       Realme  Narzo 10A   
2        2         64   4    4G VOLTE  ...       Realme  Narzo 10A   
3        2         64   4    4G VOLTE  ...       Realme  Narzo 10A   
4        2        128   4    4G VOLTE  ...       Realme   Narzo 10   

  resolution_width resolution_height f_cam_1 f_cam_2 p_cam_1 p_cam_2 p_cam_3  \
0            1600                720       5              12       2       2   
1            1600                720       5              12       2       2   
2            1600                720       5              12       2       2   
3            1600                720       5              12       2       2   
4            1600                720      16              48       8       2   

  p_cam_4  
0          
1          
2          
3          
4       2  

[5 rows x 26 columns]

In [9]:
mobile.sample(10)

mobile_price     mobile_color  dual_sim disp_size                  os  \
399      ₹60,990     Bordeaux Red  Dual Sim    16.26   Android Oreo 8.1.0   
380       ₹8,990             Blue  Dual Sim    15.75     Android Oreo 8.1   
12       ₹31,999     Neptune Blue  Dual Sim    16.51        Android Pie 9   
394      ₹12,990            Black  Dual Sim    15.75     Android Oreo 8.1   
377      ₹16,990     Marine Green  Dual Sim    16.51        Android Pie 9   
269      ₹18,900      Matte Black  Dual Sim    14.48   Android Nougat 7.1   
73        ₹9,499      Diamond Red  Dual Sim     15.8        Android Pie 9   
84       ₹16,939  More Than White  Dual Sim    15.44        Android Pie 9   
40       ₹36,999        Red Brick  Dual Sim    16.51        Android Pie 9   
310      ₹10,999    Charcoal Blue  Dual Sim    15.06     Android Oreo 8.1   

     num_cores mp_speed int_memory ram network_top  ... mobile_brand  \
399  Octa Core    2.649        256   8    4G VOLTE  ...         OPPO   
380  Octa Core      2.3         32   2    4G VOLTE  ...         OPPO   
12   Octa Core     2.96        128   8    4G VOLTE  ...       Realme   
394  Octa Core      2.3         32   2    4G VOLTE  ...         OPPO   
377  Octa Core      1.8        128   4    4G VOLTE  ...         OPPO   
269  Octa Core      1.8         32   4    4G VOLTE  ...         Vivo   
73   Octa Core      2.1         64   4    4G VOLTE  ...       Realme   
84   Octa Core        2        128   6    4G VOLTE  ...           Mi   
40   Octa Core     2.96        256  12    4G VOLTE  ...       Realme   
310  Octa Core        2         64   4    4G VOLTE  ...      Infinix   

        model_name resolution_width resolution_height f_cam_1 f_cam_2 p_cam_1  \
399         Find X            2340               1080      25              16   
380            A5s            1520                720       8              13   
12          X2 Pro            2400               1080      16              64   
394            A5s            1520                720       8              13   
377        A9 2020            1600                720      16              48   
269             V7            1440                720      24              16   
73               3            1520                720      13              13   
84              A3            1560                720      32              48   
40          X2 Pro            2400               1080      16              64   
310  Note 5 Stylus            1080               2160      16              16   

    p_cam_2 p_cam_3 p_cam_4  
399      20                  
380       2                  
12       13       8       2  
394       2                  
377       8       2       2  
269                          
73        2                  
84        8       2          
40       13       8       2  
310                          

[10 rows x 26 columns]

Last but not least, we convert the "mobile_price", from Indian Rupee into Singapore Dollar. Using the currency rate on 11 Mar 2022.

In [10]:
mobile["mobile_price"] = mobile["mobile_price"].apply(lambda x: re.sub(r'(.*)\,(.*)',r'\1\2', x))
mobile["mobile_price"] = mobile["mobile_price"].apply(lambda x: re.sub(r'.(\d+)',r'\1', x))
mobile["mobile_price"] = pd.to_numeric(mobile["mobile_price"])
mobile["mobile_price"] = mobile["mobile_price"].apply(lambda x: round(x*0.0178), x)

In [11]:
mobile.sample(10)

mobile_price     mobile_color  dual_sim disp_size                    os  \
69            587     Glacier Blue  Dual Sim    16.69             Android 10   
241           267     Starry Black  Dual Sim     15.8       Android Oreo 8.1   
141           196      Space Black  Dual Sim     16.0        Android Pie 9.0   
322           214       Topaz Blue  Dual Sim    14.35         Android Oreo 8   
84            302  More Than White  Dual Sim    15.44          Android Pie 9   
209           320      Sonic Black  Dual Sim    16.59          Android Pie 9   
267           267     Starry Black  Dual Sim     15.8       Android Oreo 8.1   
413           349      Astral Blue  Dual Sim    16.28       Android Oreo 8.1   
32            214      Crystal Red  Dual Sim    16.51          Android Pie 9   
136           231             Gold  Dual Sim    15.21   Android Nougat 7.1.2   

     num_cores mp_speed int_memory ram network_top  ... mobile_brand  \
69   Octa Core     2.96        256  12    4G VOLTE  ...       Realme   
241  Octa Core        2         64   3    4G VOLTE  ...         Vivo   
141  Octa Core        2         64   4    4G VOLTE  ...        Redmi   
322  Octa Core      1.4         64   4    4G VOLTE  ...      Infinix   
84   Octa Core        2        128   6    4G VOLTE  ...           Mi   
209  Octa Core      2.3        128   6    4G VOLTE  ...         Vivo   
267  Octa Core        2         64   3    4G VOLTE  ...         Vivo   
413  Octa Core     1.95         64   4    4G VOLTE  ...         OPPO   
32   Octa Core        2        128   4    4G VOLTE  ...       Realme   
136  Octa Core      1.8         64   4    4G VOLTE  ...        Redmi   

       model_name resolution_width resolution_height f_cam_1 f_cam_2 p_cam_1  \
69   X3 SuperZoom            2400               1080      32       8      64   
241           Y93            1520                720       8              13   
141    Note 7 Pro            2340               1080      13              48   
322        Hot S3             720               1440      20              13   
84             A3            1560                720      32              48   
209         Z1Pro            2340               1080      32              16   
267           Y93            1520                720       8              13   
413            K1            2340               1080      25              16   
32             5s             720               1600      13              48   
136    Note 5 Pro            2160               1080      20              12   

    p_cam_2 p_cam_3 p_cam_4  
69        8       8       2  
241       2                  
141       5                  
322                          
84        8       2          
209       2       8          
267       2                  
413       2                  
32        8       2       2  
136       5                  

[10 rows x 26 columns]

Finally, we rearrange our columns, to make it clearer during the interpretation process

In [12]:
list(mobile.columns.values)

['mobile_price',
 'mobile_color',
 'dual_sim',
 'disp_size',
 'os',
 'num_cores',
 'mp_speed',
 'int_memory',
 'ram',
 'network_top',
 'bluetooth',
 'battery_power',
 'mob_width',
 'mob_height',
 'mob_depth',
 'mob_weight',
 'mobile_brand',
 'model_name',
 'resolution_width',
 'resolution_height',
 'f_cam_1',
 'f_cam_2',
 'p_cam_1',
 'p_cam_2',
 'p_cam_3',
 'p_cam_4']

In [13]:
mobile = mobile[[
 'mobile_brand',
 'model_name',
 'mobile_color',
 'dual_sim',
 'disp_size',
 'os',
 'num_cores',
 'mp_speed',
 'int_memory',
 'ram',
 'network_top',
 'bluetooth',
 'battery_power',
 'mob_width',
 'mob_height',
 'mob_depth',
 'mob_weight',
 'resolution_width',
 'resolution_height',
 'f_cam_1',
 'f_cam_2',
 'p_cam_1',
 'p_cam_2',
 'p_cam_3',
 'p_cam_4',
 'mobile_price']]

In [14]:
mobile.shape

(459, 26)

Now the dataset consists of 459 rows and 26 columns. We export the cleaned dataset into a csv file to be used during the EDA process

In [15]:
mobile.to_csv("mobile_data_new.csv",mode='w+' )